src/H5Tconv_macros.h

```diff
@@ -804,6 +853,564 @@ typedef struct H5T_conv_hw_t {
+        H5T_CONV_##STYPE##_REALVAL(S); /* Extract "real" part of complex number */                           \
+        *(D) = (DT)(sr_val);                                                                                 \
+    }
+#define H5T_CONV_CAST_zX(STYPE, DTYPE, S, S_REAL, S_IMAG, D, ST, DT)                                         \
+    {                                                                                                        \
+        H5T_CONV_##STYPE##_REALVAL(S); /* Extract "real" part of complex number */                           \
+        *(D) = (DT)(sr_val);                                                                                 \
+    }
+#define H5T_CONV_CAST_fZ(STYPE, DTYPE, S, S_REAL, S_IMAG, D, ST, DT)                                         \
+    {                                                                                                        \
+        H5T_CONV_CAST_TO_##DTYPE(*(S), (ST)0.0, D, DT)                                                       \
+    }
+#define H5T_CONV_CAST_Fz(STYPE, DTYPE, S, S_REAL, S_IMAG, D, ST, DT)                                         \
+    {                                                                                                        \
+        H5T_CONV_CAST_TO_##DTYPE(*(S), (ST)0.0, D, DT)                                                       \
+    }
+#define H5T_CONV_CAST_fz(STYPE, DTYPE, S, S_REAL, S_IMAG, D, ST, DT)                                         \
+    {                                                                                                        \
+        H5T_CONV_CAST_TO_##DTYPE(*(S), (ST)0.0, D, DT)                                                       \
+    }
+#define H5T_CONV_CAST_xZ(STYPE, DTYPE, S, S_REAL, S_IMAG, D, ST, DT)                                         \
+    {                                                                                                        \
+        H5T_CONV_CAST_TO_##DTYPE(*(S), (ST)0.0, D, DT)                                                       \
+    }
+
+#define H5T_CONV_CAST_Z(SYMBOLS, STYPE, DTYPE, S, S_REAL, S_IMAG, D, ST, DT)                                 \
+    do {                                                                                                     \
+        H5T_CONV_CAST_##SYMBOLS(STYPE, DTYPE, S, S_REAL, S_IMAG, D, ST, DT)                                  \
+    } while (0)
+
+#else
+
+/* Map all complex number casts to direct casts */
+#define H5T_CONV_CAST_Z(SYMBOLS, STYPE, DTYPE, S, S_REAL, S_IMAG, D, ST, DT)                                 \
+    do {                                                                                                     \
+        *(D) = (DT)(*(S));                                                                                   \
+    } while (0)
+
+#endif
+
+#define H5T_CONV_zZ_CORE(STYPE, DTYPE, S, D, ST, DT, D_MIN, D_MAX)                                           \
```

Comment:
All of the macros below are generally identical to their floating-point counterparts (replace "Z" with "F" and "z" with "f"), but some additional logic was needed to extract and handle the real and imaginary parts as necessary. Since MSVC uses structure types for complex numbers, these macros also had to take into account that direct casts to/from complex number types can't be made during conversions when compiled with MSVC.